src/consumption.cpp

```diff
@@ -207,14 +219,27 @@ std::map<vitamin_id, int> player::vitamins_from( const item &it ) const
         return res;
     }
 
-    // food to which the player is allergic to never contains any vitamins
-    if( allergy_type( it ) != MORALE_NULL ) {
-        return res;
-    }
-
     // @todo: bionics and mutations can affect vitamin absorption
-    for( const auto &e : it.type->comestible->vitamins ) {
-        res.emplace( e.first, e.second );
+    if( it.components.size() > 0 && !it.has_flag( "NUTRIENT_OVERRIDE" ) ) {
+        // if an item is a byproduct, it should subtract the calories and vitamins instead of add
+        int byproduct_multiplier = 1;
+        for( const auto &comp : it.components ) {
```

Comment:
Taking a closer look at this it looks like it loops through components recursively and calculates the sum of their vitamin contributions. The thing I don't understand is why byproducts substract the vitamin value instead of simply ignoring it. Maybe I'm missing something but to me it seems that all you need is the sum of non-byproduct components.

As it is you can have negative vitamin values if the byproduct has high vitamin values which wouldn't make sense. Also if byproducts are ignored you can avoid recursive calls for those components by making the `comp.has_flag( "BYPRODUCT" )` check a condition instead of a multiplier.

On a different note, the `map.operator[](value)` syntax feels a bit weird, you should be able to use `map[value]` as with any overloaded operator.

Reply:
the reason is that byproducts are added to the components in the recipe, and they are subtracted here so byproducts plus actual products have values that add up to the whole

Reply:
also i have never gotten ```map[value]``` to work without giving me some kind of compile error...

Reply:
That makes sense but I don't see where the "whole" value comes from, any usages I could find assume that this function returns the calculated vitamin value of the item. I'm sorry if I'm missing something obvious but it doesn't seem to use the default vitamin values on the crafted item anywhere. It simply adds all the vitamins from the components and returns that.

There is a reference to the default vitamin values for the item but it's inside the else block, thus being ignored when the item has components. Maybe you meant to calculate the component's vitamin values and then add them to the default values?

Reply:
at line 228 it calls this->vitamins_from(comp) making it recursive, meaning the first part of the if statement will be true if there's no components, getting the default values

Reply:
Yeah, but that only works for the components so for the main item it will skip that part and only use the values returned from the components.

So if you have, lets say a steak with french fries made with meat and potatoes, it will get the vitamin values of the meat and potato components, sum them up and return that value. It won't actually get the default values for the" steak with french fries" item because that part is inside the else block for the steak item.

Reply:
Yes, that's the intent. To override the default items with the sun of its components.

Reply:
Ok, I think I know what I was missing. The byproducts aren't part of the item but the components don't reflect that so you have to substract the vitamin values from the total value. I missed the part about byproducts being different items resulting from the same components. Silly me.

It makes total sense now.